src/id_allocator.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+/// Error type for IdAllocator usage.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// All ids from the range specified are exhausted.
+    Overflow,
+    /// An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
+    /// There are no more IDs available in the manage range
+    ResourceExhausted,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            InvalidInput(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+///
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+///
+/// let irq_nr = legacy_irq_allocator.allocate_id();
+/// if irq_nr.is_err() {
+///     println!("Interrupt number allocation failed: {:?}", irq_nr);
```

Comment:
Will the test case reported as failure when `irq_nr.is_err()` is true?
Seems it will break "cargo test".